xapian-core/include/xapian/cluster.h

```diff
@@ -0,0 +1,603 @@
+
+    doccount get_doccount() const;
+};
+
+/** Abstract class representing a point in the VSM
+ */
+class XAPIAN_VISIBILITY_DEFAULT PointType
+    : public Xapian::Internal::opt_intrusive_base {
+  protected:
+    /** Implement a map to store the terms within a document
+     *  and their pre-computed TF-IDF weights
+     */
+    std::unordered_map<std::string, double> weights;
+
+    /// Store the squared magnitude of the PointType
+    double magnitude;
+
+    /** Set the weight 'weight' to the mapping of a term
+     *
+     *  @param term	Term for which the weight is supposed
+     *			to be changed
+     *  @param weight	The weight to which the mapping of the
+     *			term is to be set
+     */
+    void set_weight(const std::string &term, double weight);
+
+  public:
+    /// Constructor
+    PointType() { magnitude = 0; }
+
+    /// Return a TermIterator to the beginning of the termlist
+    TermIterator termlist_begin() const;
+
+    /// Return a TermIterator to the end of the termlist
+    TermIterator termlist_end() const {
+	return TermIterator(NULL);
+    }
+
+    /** Validate whether a certain term exists in the termlist
+     *  or not by performing a lookup operation in the existing values
+     *
+     *  @param term	Term which is to be searched
+     */
+    bool contains(const std::string &term) const;
+
+    /** Return the TF-IDF weight associated with a certain term
+     *
+     *  @param term	Term for which TF-IDF weight is returned
+     */
+    double get_weight(const std::string &term) const;
+
+    /** Add the weight 'weight' to the mapping of a term
+     *
+     *  @param term	Term to which the weight is to be added
+     *  @param value	Weight which has to be added to the existing
+     *			mapping of the term
+     */
+    void add_weight(const std::string &term, double weight);
+
+    /// Return the pre-computed squared magnitude
+    double get_magnitude() const;
+
+    /// Return the size of the termlist
+    Xapian::termcount termlist_size() const;
+
+    /** Start reference counting this object.
+     *
+     *  You can hand ownership of a dynamically allocated PointType
+     *  object to Xapian by calling release() and then passing the object to a
+     *  Xapian method.  Xapian will arrange to delete the object once it is no
+     *  longer required.
+     */
+    PointType * release() {
+	opt_intrusive_base::release();
+	return this;
+    }
+
+    /** Start reference counting this object.
+     *
+     *  You can hand ownership of a dynamically allocated PointType
+     *  object to Xapian by calling release() and then passing the object to a
+     *  Xapian method.  Xapian will arrange to delete the object once it is no
+     *  longer required.
+     */
+    const PointType * release() const {
+	opt_intrusive_base::release();
+	return this;
+    }
+};
+
+/** Class to represent a document as a point in the Vector Space
+ *  Model
+ */
+class XAPIAN_VISIBILITY_DEFAULT Point : public PointType {
```

Comment:
`PointType` is a PIMPL class, so subclassing it doesn't really work in a helpful way.  The PIMPL class is itself really just a reference-counted pointer, so the subclassing gets sliced off when you pass the subclass to things.

I'm not sure if this is actually a problem here - perhaps we only pass `Point` or `Centroid` as a `PointType` to places which only care about the `PointType` part.  It still seems a little problematic as part of the a public API though, since it makes these classes behave oddly if the user tries to use them in ways which rely on the class hierarchy behaving as it usually would.